tests/local/mod_accredible_accredible_test.php

```diff
@@ -0,0 +1,182 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace mod_accredible\local;
+
+/**
+ * Unit tests for mod/accredible/classes/local/accredible.php
+ *
+ * @package    mod_accredible
+ * @subpackage accredible
+ * @category   test
+ * @copyright  Accredible <dev@accredible.com>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mod_accredible_accredible_test extends \advanced_testcase {
+    /**
+     * @var The accredible instance.
+     */
+    protected $accredible;
+
+    /**
+     * Setup testcase.
+     */
+    public function setUp(): void {
+        $this->resetAfterTest();
+        $this->accredible = new accredible();
+    }
+
+    /**
+     * Test save_record method when creating a new record.
+     * @covers ::save_record
+     */
+    public function test_save_record_new() {
+        global $DB;
+
+        // Create a mock of the $DB object.
+        $DB = $this->createMock(\moodle_database::class);
+
+        $post = $this->generatePostObject();
+
+        // Set up the expectation for the insert_record method.
+        $DB->expects($this->once())
+            ->method('insert_record')
+            ->with('accredible', $this->anything())
+            ->willReturn(1);
+
+        $result = $this->accredible->save_record($post);
+        $this->assertEquals(1, $result);
+    }
+
+    /**
+     * Test save_record method when updating an existing record.
+     * @covers ::save_record
+     */
+    public function test_save_record_update() {
+        global $DB;
+
+        // Create a mock of the $DB object.
+        $DB = $this->createMock(\moodle_database::class);
+
+        $overrides = new \stdClass();
+        $overrides->name = 'Updated Certificate';
+        $overrides->instance = 1;
+        $post = $this->generatePostObject($overrides);
+
+        $accrediblecertificate = new \stdClass();
+        $accrediblecertificate->achievementid = null;
+
+        // Simulating update_record return value.
+        $DB->expects($this->once())
+            ->method('update_record')
+            ->with('accredible', $this->anything())
+            ->willReturn(true);
+
+        $result = $this->accredible->save_record($post, $accrediblecertificate);
+        $this->assertTrue($result);
+    }
+
+    /**
+     * Test save_record method when attribute mappings are empty.
+     * @covers ::save_record
+     */
+    public function test_attributemapping_is_null_when_mappings_are_empty() {
```

Comment:
I think as the existing pattern in this plugin code, we create one test function per function with the prefix `test_[function name]` and include test scenarios in the same function.

So if we want to test `save_record`, I think we should include all test scenarios for the function in `test_save_record`.

Please correct me if my understanding is outdated.

Reply:
I was following the pattern we use on the FE. I'll refactor the tests so we have all the scenarios in a single function.